### PR TITLE
Auto mixed precision conversion of GPT-2 onnx model

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**:

Add an option "--auto_mixed_precision" to convert GPT-2 model from float32 to mixed precision. It will detect whether the weights are float16. If so, the output model will have all weights converted to float16, and logits output is also float16. Otherwise, weights for nodes computed in float32 will still remain float32, and logits will be float32.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Automatic converting fp32 model to mixed precision, with good tradeoff between precision and performance.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/convert_to_onnx.py [hunk 1]
```diff
@@ -159,6 +159,14 @@ def parse_arguments(argv=None):
     fp16_option_group = parser.add_argument_group(
         "float to float16 conversion parameters that works when \"--precision fp16\" is specified")
 
+    fp16_option_group.add_argument(
+        '-a',
+        '--auto_mixed_precision',
+        required=False,
+        action='store_true',
+        help='Convert to mixed precision automatically. Other float16 conversion parameters will be ignored.')
```
Comment (viboga) [inline onnxruntime/python/tools/transformers/convert_to_onnx.py:167]: Is there a reason to allow other parameters for float16? Any fallback mechanism uses those parameters?

File: onnxruntime/python/tools/transformers/gpt2_helper.py [hunk 2]
```diff
@@ -501,13 +578,15 @@ def onnxruntime_inference(ort_session, inputs: Gpt2Inputs, total_runs: int = 0):
...
-        return IOBindingHelper.prepare_io_binding(ort_session, input_ids, position_ids, attention_mask, past, output_buffers, output_shapes)
+        return IOBindingHelper.prepare_io_binding(ort_session, input_ids, position_ids, attention_mask, past,
+                                                  output_buffers, output_shapes)
 
     @staticmethod
     def get_outputs_from_io_binding_buffer(ort_session, output_buffers, output_shapes, return_numpy=True):
         """ Copy results to cpu. Returns a list of numpy array.
         """
-        return IOBindingHelper.get_outputs_from_io_binding_buffer(ort_session, output_buffers, output_shapes, return_numpy)
+        return IOBindingHelper.get_outputs_from_io_binding_buffer(ort_session, output_buffers, output_shapes,
```
Comment (viboga) [inline onnxruntime/python/tools/transformers/gpt2_helper.py:588]: I think its a good idea to handle refactoring as diff PR.
It basically makes the change look bigger than it actually is.

File: onnxruntime/python/tools/transformers/fusion_utils.py [hunk 3]
```diff
@@ -123,6 +124,24 @@ def get_dtype(self, shape_infer_helper, input_or_output_name: str) -> int:
 
         return None
 
+    def remove_cascaded_cast_nodes(self):
+        """Remove Cast node that are overrided by another Cast node like  --> Cast --> Cast -->
+           Note that this shall be used carefully since it might introduce semantic change.
+           For example, float -> int -> float could get different value than the original float value.
```
Comment (viboga) [inline onnxruntime/python/tools/transformers/fusion_utils.py:130]: I see this is getting called only once and first part of this comment is taken care of. 
Is the second part 
'float -> int -> float could get different value than the original float value.' taken care somewhere else?

File: onnxruntime/python/tools/transformers/convert_to_onnx.py (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/convert_to_onnx.py:167]: The fallback mechanism: when use opt to auto, we will generate all parameters automatically. Otherwise, use has to specify other parameters manually.

File: onnxruntime/python/tools/transformers/fusion_utils.py (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/fusion_utils.py:130]: Mixed precision conversion adds some "Cast" nodes to the graph. Here we have a post processing to remove some "Cast" nodes that is not needed.

File: onnxruntime/python/tools/transformers/gpt2_helper.py (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/gpt2_helper.py:588]: This line is changed by format tool (yapf) automatically.

**Review by viboga — APPROVED** [review]
